### ISSUE TITLE
File having been compared is locked, cannot be cleaned up

### ISSUE THREAD

**Issue by oliver-gramberg** [manykarim/robotframework-doctestlibrary#17]
My test downloads a PDF to a temporary local folder, then runs keyword `Compare Images` on that file as `test_img` and a reference file as `reference_img`.

Attempting to `Empty Directory    ${download directory}` on that local folder during the following test, or to delete it and all its contents with `Remove Directory        ${download directory}   recursive=True` during suite teardown both fail:

PermissionError: [WinError 32] Process cannot access this file because it is used by another process: '<path to PDF omitted>'

(Translation by me, so may be different in the English original.)

I guess that some part of the image processing is still active/has not closed the file. Any hints?

**Comment by manykarim** [comment]
Thanks a lot for your finding Oliver, I'll look into it and check if the library keeps a lock on the document files - even though I cannot really imagine why.
One thing you could maybe try:
If you delete the temporary folder in your automated test without executing the Compare Images keyword - is the deletion successful? Just to ensure that Compare Images is really causing it

**Comment by oliver-gramberg** [comment]
> If you delete the temporary folder in your automated test without executing the Compare Images keyword

I did try exactly that, and yes it deletes without problems.

**Comment by podfran** [comment]
I have the same issue. Is there any progress?

**Comment by podfran** [comment]
After some investigation I have found two things:
1. This issue only seems to apply when a file comparison fails. The file is locked until the entire program exits, so any subsequent test setup/teardown are unable to manipulate the file. Neither can the file be removed from RF or externally while the program is still running.
2. It seems that there are some unclosed threads from the `CompareImages` class, which do not close when a test finishes. The program seems to hang for about 30s-1m before exiting finally after RF displays the end results of the run in the console, as if the program was waiting for some threads and then killing them after some timeout.

**Comment by manykarim** [comment]
Thanks a lot for your additional investigation @podfran .
Damn, now I have an idea.

I use the `futures.ThreadPoolExecutor` multiple times in the library for multithreading.
Mostly, to parallelize processing of reference/candidate image at the same time or comparing multiple pages at once (for multi-page documents).
https://github.com/manykarim/robotframework-doctestlibrary/blob/main/DocTest/VisualTest.py#L209

What you now describe sounds like those threads are hanging.

**Comment by podfran** [comment]
Having a debug through the `VisualTest` library, the number of threads at the eand of each passing and failing paths was the same. This onlytells me that the unclosed file handle is somewhere in the main thread. Note that it is closed correctly when the comparison passes. No issues with manipulating the candidate file after `compare images`.

BTW, this is an issue I have tested across versions of `DocTestLibrary`, not only in the latest version. 

Also, for clarity's sake, I'm comparing PDF files. Maybe this issue does not apply to other file formats. Haven't checked.

**Comment by ValentinoAlfano** [comment]
I have the same issue when the `Compare Pdf Documents` keyword fails.

This is particularly annoying because once that keyword has failed, files used as arguments can't be used anymore throughout that RF test execution, since I think there is no way to hijack and close the hanging process without killing python & RF theirself.

Are there any updates about a fix, even a workaround or temporary one?

Thank you

**Comment by manykarim** [comment]
Thank you @ValentinoAlfano , I think I finally found something.
This issue from pymupdf sounded familiar https://github.com/pymupdf/PyMuPDF/discussions/1547
In some cases pymupdf keeps a lock on the files.

I solved it by setting the pymupdfdoc to `None` before the Assertion Error is thrown.

```
      reference_compare_image.mupdfdoc = None
      candidate_compare_image.mupdfdoc = None
``` 

I changed the following in  VisualTest.py in lines  248/249

```
  if (difference):
      print("The compared images are different")
      reference_compare_image.mupdfdoc = None
      candidate_compare_image.mupdfdoc = None
      raise AssertionError('The compared images are different.')
``` 

I tried locally and the .PDF files (which were locked before in failed comparisons) were released.

Can you give it a quick yourself and modify the file once and check if it also fixes the problem for you?
I will then add a fix and release it officially.

Sorry that it took so long

**Comment by manykarim** [comment]
ok, I released a new version 0.12.0 with that fix.
Would be great if you can give it a try and let me know the outcome :)

**Comment by ValentinoAlfano** [comment]
@manykarim  thank you I just tried your suggestion into the compare_pdf_documents():

```
if differences_detected:
	ref_doc = None
	cand_doc = None
	raise AssertionError('The compared PDF Document Data is different.')
```

And I think this solved the problem there too (I performed three runs without errors on accessing the files immediately after the keyword failure).

Thank you so much!

**Comment by ValentinoAlfano** [comment]
> ok, I released a new version 0.12.0 with that fix. Would be great if you can give it a try and let me know the outcome :)

I think your fix didn't affect the PdfTest.py keywords, that source file is still at a commit from July 2022.

**Comment by manykarim** [comment]
Right, I also need to add it to the PDF document comparison. Thanks for checking that as well !!

**Comment by manykarim** [comment]
ALright, just published 0.12.1 which also includes the fix for PdfTests

**Comment by ValentinoAlfano** [comment]
Hi @manykarim 

I know this issue has been closed, but I noticed that when Compare Pdf Documents fails for FileNotFoundError (in my case reference file path was not found), candidate pdf file (first argument) is still not unlocked.

**Comment by manykarim** [comment]
Oh, nice 🪲 @ValentinoAlfano 
thank you, I will re-open the issue and fix it in the next days
